### ISSUE TITLE
Behave tests hang

### ISSUE THREAD

**Issue by rasmuse** [demokratikollen/demokratikollen#63]
Behave tests are hanging for me most of the time. Sometimes one or two of the tests run and give reasonable output, but most of the time the application hangs. By inserting some debugging output here and there I have concluded that the web page is not being delivered to the PhantomJS client.

I have just gone through the following steps:
1. Check out latest commit on `develop`, e246a9766290ad0924a265a7b3128c6bf99eb86e
2. `vagrant destroy` followed by `vagrant up`
3. Then, on the virtual machine,

``` bash
(python3.4) vagrant@vagrant-ubuntu-trusty-32:~/demokratikollen/www$ behave
Starting Webserver and Webdriver
/home/vagrant/.virtualenv/python3.4/lib/python3.4/site-packages/flask_cache/__init__.py:152: UserWarning: Flask-Cache: CACHE_TYPE is set to null, caching is effectively disabled.
  warnings.warn("Flask-Cache: CACHE_TYPE is set to null, "
Feature: Test the contact static page. # features/static.contact.feature:1

  Scenario: There should be a contact page                                  # features/static.contact.feature:3
    Given you browse to the contact page                                    # features/steps/static.py:9
```

At this point, the application hangs, it seems forever. Killing with Ctrl+C and then running again gave the same (lack of) response.
1. `vagrant halt` followed by `vagrant up`
2. Testing again:

``` bash
(python3.4) vagrant@vagrant-ubuntu-trusty-32:~/demokratikollen/www$ behave
Starting Webserver and Webdriver
Feature: Test the contact static page. # features/static.contact.feature:1

  Scenario: There should be a contact page                                  # features/static.contact.feature:3
    Given you browse to the contact page                                    # features/steps/static.py:9
/home/vagrant/.virtualenv/python3.4/lib/python3.4/site-packages/flask_cache/__init__.py:152: UserWarning: Flask-Cache: CACHE_TYPE is set to null, caching is effectively disabled.
    Given you browse to the contact page                                    # features/steps/static.py:9 0.889s
    Then The response be a page with the title "Demokratikollen - Kontakt"! # features/steps/static.py:14 0.012s

  Scenario: The correct header_button should be activated                   # features/static.contact.feature:7
    Given you browse to the contact page                                    # features/steps/static.py:9 0.181s
    Then The response should have only the "Kontakt" page button activated. # features/steps/static.py:18 0.642s

Feature: Test the index static page. # features/static.index.feature:1

  Scenario: There should be an index page                               # features/static.index.feature:3
    Given you browse to the index page                                  # features/steps/static.py:4 0.147s
    Then The response be a page with the title "Demokratikollen - Hem"! # features/steps/static.py:14 0.009s

  Scenario: The correct header_button should be activated               # features/static.index.feature:7
      Given you browse to the index page                                  # features/steps/static.py:4 35.539s
```

So now it suddenly works for a few tests, then hangs again.

Does @joharohl or anyone else understand what's going on?

**Comment by joharohl** [comment]
I had similar issues previously. I thought I solved it. 

If it is the same issue the reason for the bug is that the webserver is spawned off as a independent process in the behave setup file (`features/environment.py`). Unfortunately, the startup code continues to execute immediately (it sort of has to) and then there is a race condition. The best solution would of course be to have a test at the end of the setup to make sure that the server is up and running, but for now, a time delay should do it. I just moved the startup of the server before the PhantomJS startup and that was enough for me. Maybe your machine is faster.

I added a sleep, check out d35952542205d2c8836e66c158cbe0b30e695cd0 and see if it helps.

PS. use `ps -ef | grep gunicorn` if it gets stuck to get the PID (second column) of the server if behave hangs. That way you can kill it with `kill PID`.

**Comment by rasmuse** [comment]
Thanks! Unfortunately does not work. And your theory doesn't explain the cases where it manages to run a few tests and then crashes, right?

Now, I came up with a couple of ideas in the mean time, and maybe you can figure something out.

First, I have been in a google frenzy, and it seems like other projects also have had problems with PhantomJS locking their applications. One theory I've seen is that a single-threaded server (gunicorn is?) gets problems if the client doesn't close the TCP socket properly, because then someone can end up waiting for a message that never comes, and the server locks. Not sure about this, but in any case I tried running the tests with `context.driver = webdriver.Firefox()` instead, and it successfully finishes all the tests every time. Conclusion: This mess has something to do with browser behavior.

To run Firefox headless, you need a dummy "display" connected to Firefox. First do `sudo apt-get install firefox` and then install xvfb as explained here: http://www.installationpage.com/selenium/how-to-run-selenium-headless-firefox-in-ubuntu/

I start two ssh sessions: In the first, i run the xvfb server (`sudo Xvfb :10 -ac`). In the second, I first `export DISPLAY=:10` and then run `behave`.

---

My second idea was, that maybe we are not actually using the gunicorn server at all here, since all the tests are serving static content? I tried removing everything with gunicorn so I have

``` Python
def before_all(context):
   context.driver = webdriver.Firefox()

def after_all(context):
    pass
```

running `behave` with this setup works fine. All the tests finish in about 17 seconds.

Now please enlighten me :)

**Comment by joharohl** [comment]
True about hanging midway. Missed that. 

I picked PhantomJS since it was headless. Found about the way to run Firefox (or others) headless after a while but never bothered switching since PhantomJS worked for me. 

It is a good idea to switch to Firefox anyway since PhantomJS has some other issues with javascript and what not. 

Running behave and still getting responses is weird indeed. The content needs to be served to Firefox somehow. If it is truly static content you might get away with browsing the files with the filepath but I would guess a lot of dependencies would break (like js and css files). Unless something magic happens when one starts the webdrive,  I would say a server is still running somewhere. How does your tests look like? Does `ps -ef | grep gunicorn` show anything?

**Comment by rasmuse** [comment]
Well, I thought that `ps -ef | grep gunicorn` showed nothing, but now I ran it again and it obviously doesn't work without a server in the `before_all` and `after_all` functions. So then it makes sense that we have some bad network code in the PhantomJS client which causes the conversation to hang.

So let's switch to Chrome or Firefox or some other browser in the behave tests. But this issue remains, I guess? Our production site will also run ngnix/gunicorn, right? Is there anything telling us that people with bad browser won't freeze our server threads in the production environment? (I might be missing something, because I barely understand all the black magic we have installed now...)

**Comment by rasmuse** [comment]
More info!

Just upgraded to phantomjs 1.9.8 (from 1.9.0 which is the one we get with `apt-get`). Tests run fast (about 2.5 seconds on my machine) and never seem to hang. If this will do for the behave tests, I'd rather keep phantomjs because it speeds up the testing cycle significantly.

@joharohl, would you be kind and set us up with a better browser than phantomjs 1.9.0 in the provisioning? My suggestion for now is phantomjs 1.9.8, but I guess other things will work too.

My previous question still remains though: Will we be able to choke the production environment by connecting with a bunch of bad browsers (phantomjs 1.9.0 for example)?

**Comment by joharohl** [comment]
Great news! I'll touch up the provisioning. Unless we specifically need a "proper" browser I think that PhanomJS is a good option as long as it works.

Short answer to your question: I don't think so. Long one:

The Gunicorn server consists of a single "master" process and several worker processes. The master process monitors the worker processes while the worker processes handles incoming requests. Usually, a few worker processes are enough since the time for a single process is relatively small and even for a large site the number of requests per second is less than 100 (many users but they don't browse to a new page that often). 

There are timeouts concerning connections that do not close properly. Default settings are to restart the worker if an open connection is silent for more than two seconds. In the behave case this is enough to fudge its operation to the point where it hangs. I am surprised it doesn't retry after a few second.

Anyhow, on the production site we run `num_cpus*2 + 1` workers as per gunicorns recommendations. In practice, only a deliberate attacker might block all of these workers. In the worst case, users might have to reload the page, two seconds is not that long. If we implement proper caching this risk would be even lower. My guess is that if there is a concern for attackers you would have some sort of network filtering before gunicorn, either in nginx or a dedicated software (if the request reaches gunicorn the damage is already done).

**Comment by joharohl** [comment]
And I forgot: On the dev VM gunicorn is run using a single worker process, leading to the behave issues.

**Comment by rasmuse** [comment]
Thanks. I just spent a couple of hours reading up on this stuff instead of working, so now I understand much better now.

I have also been experimenting a bit back and forth with the tests. The gunicorn worker thread still hangs sometimes, but eventually it reboots and tests complete successfully.

Please close this issue when you have put PhantomJS 1.9.8 in provisioning.

**Comment by joharohl** [comment]
Should be fixed since 84753ce9bef38f40471f87e58407c81f753ff15e. Uninstall the package `phantomjs` on the Vm and run a `vagrant up` to install newest phantomjs.
